File: flytekit-examples/src/main/java/org/flyte/examples/WelcomeWorkflow.java
/*
 * Copyright 2020-2023 Flyte Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.flyte.examples;

import com.google.auto.service.AutoService;
import com.google.auto.value.AutoValue;
import org.flyte.flytekit.SdkBindingData;
import org.flyte.flytekit.SdkWorkflow;
import org.flyte.flytekit.SdkWorkflowBuilder;
import org.flyte.flytekit.jackson.Description;
import org.flyte.flytekit.jackson.JacksonSdkType;

/** Example workflow that takes a name and outputs a welcome message. */
@AutoService(SdkWorkflow.class)
public class WelcomeWorkflow extends SdkWorkflow<WelcomeWorkflow.Input, WelcomeWorkflow.Output> {

  @AutoValue
  public abstract static class Input {
    @Description("Name to be welcomed")
    public abstract SdkBindingData<String> name();

    public static WelcomeWorkflow.Input create(SdkBindingData<String> name) {
      return new AutoValue_WelcomeWorkflow_Input(name);
    }
  }

  @AutoValue
  public abstract static class Output {
    public abstract SdkBindingData<String> greeting();

    public static Output create(SdkBindingData<String> greeting) {
      return new AutoValue_WelcomeWorkflow_Output(greeting);
    }
  }

  public WelcomeWorkflow() {
    super(
        JacksonSdkType.of(WelcomeWorkflow.Input.class),
        JacksonSdkType.of(WelcomeWorkflow.Output.class));
  }

  @Override
  public WelcomeWorkflow.Output expand(SdkWorkflowBuilder builder, Input input) {
    // uses the workflow input as the task input of the GreetTask
    SdkBindingData<String> greeting =
        builder
            .apply("greet", new GreetTask(), GreetTask.Input.create(input.name()))
            .getOutputs()
            .greeting();

    // uses the output of the GreetTask as the task input of the AddQuestionTask
    SdkBindingData<String> greetingWithQuestion =
        builder
            .apply("add-question", new AddQuestionTask(), AddQuestionTask.Input.create(greeting))
            .getOutputs()
            .greeting();

    return WelcomeWorkflow.Output.create(greetingWithQuestion);
  }
}

File: flytekit-examples-scala/src/main/scala/org/flyte/examples/flytekitscala/WelcomeWorkflow.scala
/*
 * Copyright 2021-2023 Flyte Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.flyte.examples.flytekitscala

import org.flyte.flytekit.SdkBindingData
import org.flyte.flytekitscala.{
  SdkScalaType,
  SdkScalaWorkflow,
  SdkScalaWorkflowBuilder
}

/** Example workflow that takes a name and outputs a welcome message
  * |  start of workflow  |
  * |:-------------------:|
  * | input: name(string) |
  * |
  * |
  * +--------------v-----------------+
  * | GreetTask                |
  * |:-------------------------|
  * | input: name(string)      |
  * | output: greeting(string) |
  * |
  * |
  * +--------------v-----------------+
  * |     AddQuestionTask      |
  * |:------------------------:|
  * | input: greeting(string)  |
  * | output: greeting(string) |
  * |
  * |
  * +--------------v-----------------+
  * | end of workflow          |
  * |:-------------------------|
  * | output: greeting(string) |
  */

case class WelcomeWorkflowInput(name: SdkBindingData[String])
case class WelcomeWorkflowOutput(greeting: SdkBindingData[String])

class WelcomeWorkflow
    extends SdkScalaWorkflow[WelcomeWorkflowInput, WelcomeWorkflowOutput](
      SdkScalaType[WelcomeWorkflowInput],
      SdkScalaType[WelcomeWorkflowOutput]
    ) {

  override def expand(
      builder: SdkScalaWorkflowBuilder,
      input: WelcomeWorkflowInput
  ): WelcomeWorkflowOutput = {
    // defines the input of the workflow

    // uses the workflow input as the task input of the GreetTask
    val greeting = builder
      .apply("greet", new GreetTask(), GreetTaskInput(input.name))
      .getOutputs
      .greeting

    // uses the output of the GreetTask as the task input of the AddQuestionTask
    val greetingWithQuestion = builder
      .apply(
        "add-question",
        new AddQuestionTask(),
        AddQuestionTaskInput(greeting)
      )
      .getOutputs
      .greeting

    WelcomeWorkflowOutput(greetingWithQuestion)
  }
}
